File: apps/gui/spikes/tauri/src/composer.ts
export type ActionRecorder = (action: string) => void;

export class ComposerModel {
  private composing = false;
  private composition = "";
  private value = "";

  constructor(private readonly record: ActionRecorder) {}

  get draft(): string {
    return this.value;
  }

  beginComposition(): void {
    this.composing = true;
    this.composition = "";
    this.record("composition:start");
  }

  updateComposition(value: string): void {
    this.composition = value;
    this.record(`composition:update:${value}`);
  }

  commitComposition(): void {
    this.value += this.composition;
    this.record(`composition:commit:${this.composition}`);
    this.composition = "";
    this.composing = false;
  }

  paste(value: string): void {
    this.value += value;
    this.record(`paste:${value.replaceAll("\n", "\\n")}`);
  }

  submit(): boolean {
    if (this.composing) {
      return false;
    }
    this.record(`submit:${this.value.replaceAll("\n", "\\n")}`);
    return true;
  }
}

File: apps/gui/spikes/gpui/src/composer.rs
use crate::app::ActionRecorder;

pub struct ComposerModel {
    composing: bool,
    composition: String,
    value: String,
    recorder: ActionRecorder,
}

impl ComposerModel {
    pub(crate) fn new(recorder: ActionRecorder) -> Self {
        Self {
            composing: false,
            composition: String::new(),
            value: String::new(),
            recorder,
        }
    }

    pub fn draft(&self) -> &str {
        &self.value
    }

    pub fn begin_composition(&mut self) {
        self.composing = true;
        self.composition.clear();
        self.recorder.record("composition:start");
    }

    pub fn update_composition(&mut self, value: &str) {
        value.clone_into(&mut self.composition);
        self.recorder.record(format!("composition:update:{value}"));
    }

    pub fn commit_composition(&mut self) {
        self.value.push_str(&self.composition);
        self.recorder
            .record(format!("composition:commit:{}", self.composition));
        self.composition.clear();
        self.composing = false;
    }

    pub fn paste(&mut self, value: &str) {
        self.value.push_str(value);
        self.recorder
            .record(format!("paste:{}", value.replace('\n', "\\n")));
    }

    pub fn sync_from_framework(&mut self, value: &str) {
        value.clone_into(&mut self.value);
    }

    pub fn submit(&mut self) -> bool {
        if self.composing {
            return false;
        }
        self.recorder
            .record(format!("submit:{}", self.value.replace('\n', "\\n")));
        true
    }
}
